ReactiveCocoa/Swift/Signal.swift

```diff
@@ -410,16 +410,127 @@ public func takeUntil<T, E>(trigger: Signal<(), NoError>)(signal: Signal<T, E>)
 	}
 }
 
+/// Forwards events from `signal` with history: values of the returned signal
+/// are a tuple whose first member is the previous value and whose second member
+/// is the current value. `initial` is supplied as the first member when `signal`
+/// sends its first value.
+public func combinePrevious<T, E>(initial: T)(signal: Signal<T, E>) -> Signal<(T, T), E> {
+	return signal |> reduce((initial, initial)) { previousCombinedValues, newValue in
+		return (previousCombinedValues.1, newValue)
+	}
+}
+
+/// Like `scan`, but sends only the final value and then immediately completes.
+public func reduce<T, U, E>(initial: U, combine: (U, T) -> U)(signal: Signal<T, E>) -> Signal<U, E> {
+	return signal
+		|> scan(initial, combine)
```

Comment:
Ah, right. Need SignalProducer.concat for that, which is still TODO. Will try to do this evening.
> On Jan 20, 2015, at 10:26 AM, Justin Spahr-Summers <notifications@github.com> wrote:
> 
> In ReactiveCocoa/Swift/Signal.swift <https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1688#discussion-diff-23242906>:
> 
> > @@ -410,16 +410,127 @@ public func takeUntil<T, E>(trigger: Signal<(), NoError>)(signal: Signal<T, E>)
> >  	}
> >  }
> >  
> > +/// Forwards events from `signal` with history: values of the returned signal
> > +/// are a tuple whose first member is the previous value and whose second member
> > +/// is the current value. `initial` is supplied as the first member when `signal`
> > +/// sends its first value.
> > +public func combinePrevious<T, E>(initial: T)(signal: Signal<T, E>) -> Signal<(T, T), E> {
> > +	return signal |> reduce((initial, initial)) { previousCombinedValues, newValue in
> > +		return (previousCombinedValues.1, newValue)
> > +	}
> > +}
> > +
> > +/// Like `scan`, but sends only the final value and then immediately completes.
> > +public func reduce<T, U, E>(initial: U, combine: (U, T) -> U)(signal: Signal<T, E>) -> Signal<U, E> {
> > +	return signal
> > +		|> scan(initial, combine)
> There's an edge case here where the input signal is empty, so scan never sends a value. reduce should still send exactly one value (the initial value) in that case.
> 
> You could pehaps accomplish this by prepending a signal of initial before the takeLast(1).
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1688/files#r23242906>.
>